gflownet/utils/batch.py

```diff
@@ -0,0 +1,179 @@
+
+    def add_to_batch(self, envs, actions, valids, train=True):
+        for env, action, valid in zip(envs, actions, valids):
+            self.envs.update({env.id: env})
+            if not valid:
+                continue
+            if train:
+                self.state.append(deepcopy(env.state))
+                self.action.append(action)
+                self.env_id.append(env.id)
+                self.done.append(env.done)
+                self.step.append(env.n_actions)
+                mask_f = env.get_mask_invalid_actions_forward()
+                self.mask_invalid_actions_forward.append(mask_f)
+                if self.loss == "flowmatch":
+                    parents, parents_a = env.get_parents(action=action)
+                    assert (
+                        action in parents_a
+                    ), f"""
+                    Sampled action is not in the list of valid actions from parents.
+                    \nState:\n{env.state}\nAction:\n{action}
+                    """
+                    self.parents.append(parents)
+                    self.parents_actions.append(parents_a)
+                if self.loss == "trajectorybalance":
+                    mask_b = env.get_mask_invalid_actions_backward(
+                        env.state, env.done, [action]
+                    )
+                    self.mask_invalid_actions_backward.append(mask_b)
+            else:
+                if env.done:
+                    self.state.append(env.state)
+                    self.env_id.append(env.id)
+                    self.step.append(env.n_actions)
+
+    def process_batch(self):
+        self._process_states()
+        self.env_id = tlong(self.env_id, device=self.device)
+        self.step = tlong(self.step, device=self.device)
+        self._process_trajectory_indices()
+        if len(self.action) > 0:
+            self.action = tfloat(self.action, device=self.device, float=self.float)
+            self.done = tbool(self.done, device=self.device)
+            self.mask_invalid_actions_forward = tbool(
+                self.mask_invalid_actions_forward, device=self.device
+            )
+            if self.loss == "flowmatch":
+                self.parents_state_idx = tlong(
+                    sum([[idx] * len(p) for idx, p in enumerate(self.parents)], []),
+                    device=self.device,
+                )
+                self.parents_actions = torch.cat(
+                    [
+                        tfloat(x, device=self.device, float=self.float)
+                        for x in self.parents_actions
+                    ]
+                )
+            elif self.loss == "trajectorybalance":
+                self.mask_invalid_actions_backward = tbool(
+                    self.mask_invalid_actions_backward, device=self.device
+                )
+            self._process_parents()
+
+    def _process_states(self):
+        self.state_gfn = tfloat(self.state, device=self.device, float=self.float)
```

Comment:
1) `state_gfn` is a slightly confusing name (maybe `state_tensor` would be better?), 2) I would suggest initializing all of the class variables in init (this also applies to `self.trajectory_indices` etc).

Reply:
`state_gfn` means that they are in the gfn format, but these will be changed with the Alex's suggestion we discussed in the gflownet-repo channel. About initialising all the class variables in the init, could you elaborate why you think it should be this way?